iox_query/src/provider/adapter.rs

```
@@ -322,23 +359,32 @@ mod tests {
             Field::new("e", DataType::Float64, true),
             Field::new("b", DataType::Int32, false),
             Field::new("d", DataType::Float32, true),
+            Field::new("f", DataType::Utf8, true),
             Field::new("a", DataType::Int32, false),
         ]));
         let input_stream = stream_from_batch(batch.schema(), batch);
-        let adapter_stream =
-            SchemaAdapterStream::try_new(input_stream, output_schema, baseline_metrics()).unwrap();
+        let adapter_stream = SchemaAdapterStream::try_new(
+            input_stream,
+            output_schema,
+            &HashMap::from([
+                ("a", ScalarValue::from(1i32)),
+                ("f", ScalarValue::from("xxx")),
+            ]),
+            baseline_metrics(),
+        )
+        .unwrap();
 
         let output = collect(Box::pin(adapter_stream))
             .await
             .expect("Running plan");
         let expected = vec![
-            "+-----+---+---+---+---+",
-            "| c   | e | b | d | a |",
-            "+-----+---+---+---+---+",
-            "| foo |   | 4 |   | 1 |",
-            "| bar |   | 5 |   | 2 |",
-            "| baz |   | 6 |   | 3 |",
-            "+-----+---+---+---+---+",
+            "+-----+---+---+---+-----+---+",
+            "| c   | e | b | d | f   | a |",
+            "+-----+---+---+---+-----+---+",
+            "| foo |   | 4 |   | xxx | 1 |",
+            "| bar |   | 5 |   | xxx | 2 |",
+            "| baz |   | 6 |   | xxx | 3 |",
+            "+-----+---+---+---+-----+---+",
```

So if the virtual column is both in the input stream and declared to be a virtual column, then the real input column wins and is in the output.

Based on my (admittedly not deep) understanding from your notes in #6098 it seems like maybe this should be an error instead? Or is it possible that the schema adapter is applied twice, and you want the adapter to be idempotent?

---

I suggest the schema adapter should throw an error if the virtual column tries to overrride an actual column. It seems like silently picking one or the other might potentially mask bugs, and make debugging them challenging

But like @wolffcm  I may not appreciate the entire subtlety

---

Good call. I've added an error. While doing so I've also found that the schema check for NULL columns was pointless (it checked the value that it just copied) but that it missed to check if field is actually nullable. I've fixed and tested that as well.